src/librustc_span/source_map.rs

```diff
@@ -682,15 +682,7 @@ impl SourceMap {
     ///
     /// If `sp` points to `"let mut x"`, then a span pointing at `"let "` will be returned.
     pub fn span_until_non_whitespace(&self, sp: Span) -> Span {
-        let mut whitespace_found = false;
-
-        self.span_take_while(sp, |c| {
-            if !whitespace_found && c.is_whitespace() {
-                whitespace_found = true;
-            }
-
-            if whitespace_found && !c.is_whitespace() { false } else { true }
-        })
+        self.span_take_while(sp, |_| true)
     }
```

Comment:
This was a strange journey, I'll try to explain how I reduced this:

I started with the lower line
````rust
if whitespace_found && !c.is_whitespace() { false } else { true }
````
First, we swap `true` and `false` and negate the condition:
````rust
!(whitespace_found && !c.is_whitespace() { true } else { false }
````
We can remove the branches now:
````rust
!(whitespace_found && !c.is_whitespace() 
````
and lower the outer negation
````rust
!whitespace_found || c.is_whitespace()
````

which gets us
````rust
let mut whitespace_found = false;

self.span_take_while(sp, |c| {
    if !whitespace_found && c.is_whitespace() {
        whitespace_found = true;
    }

    !whitespace_found || c.is_whitespace()
})
````
The function will always return true if `c.is_whitespace()` is true, so we can put an early return inside the `if` block as a shortcut:
````rust
let mut whitespace_found = false;

self.span_take_while(sp, |c| {
    if !whitespace_found && c.is_whitespace() {
        whitespace_found = true; // dead store, remove
	return true;
    }

 !whitespace_found || c.is_whitespace()
````
now `whitespace_found = true;` is a dead store. Remove it and make `whitespace_found` no longer be mutable because it doesn't need to be
````rust
let whitespace_found = false;

self.span_take_while(sp, |c| {
    if !whitespace_found && c.is_whitespace() {
	return true;
    }

    !whitespace_found || c.is_whitespace()
````
We can inline `whitespace_found` which is now a constant
````rust
let whitespace_found = false;

self.span_take_while(sp, |c| {
    if true && c.is_whitespace() { // now we can easily see that this part always returns true
        return true;
    }

    true || c.is_whitespace()
}) 
````

as you can see, the closure should always return `true`

This seems very strange and I would like additional pairs of eyes on this!